### DISCUSSION TITLE
What is EdgeDB's prefer schema of representing interaction data?

### DISCUSSION THREAD

**Discussion by NatureGeorge** [geldata/gel#2088] [Q&A]
Hi! I am new to this awesome DB and delighted to migrate some of my projects to it. The problem I encountered is that some of my raw data are stored in the interaction table with some properties, i.e friendship table like this: 

<table>
 <tr>
  <td>Partner1</td>
  <td>Partner2</td>
  <td>InitYear</td>
 </tr>
 <tr>
  <td>Alice</td>
  <td>Bob</td>
  <td>2010</td>
 </tr>
 <tr>
  <td>Cindy</td>
  <td>Alice</td>
  <td>2007</td>
 </tr>
 <tr>
  <td>Bob</td>
  <td>Mike</td>
  <td>2018</td>
  </tr>
 <tr>
  <td>Cindy</td>
  <td>Victor</td>
  <td>2018</td>
  </tr>
<table>

> *`InitYear`: the year that they became friends; `Partner1` and `Partner2` do not require to be ordered/sorted in this case

From the tutorial, I got the idea to use the `multi link` & `expression aliases` to represent a relationship. But I cannot find a way to store the property of a particular relationship/interaction. 

Besides, what if there are some properties that are partner-specific? Like this:

<table>
 <tr>
  <td>Partner1</td>
  <td>Partner2</td>
  <td>InitYear</td>
  <td>Sent1</td>
  <td>Sent2</td>
 </tr>
 <tr>
  <td>Alice</td>
  <td>Bob</td>
  <td>2010</td>
  <td>6</td>
  <td>7</td>
 </tr>
 <tr>
  <td>Cindy</td>
  <td>Alice</td>
  <td>2007</td>
  <td>4</td>
  <td>1</td>
 </tr>
 <tr>
  <td>Bob</td>
  <td>Mike</td>
  <td>2018</td>
  <td>11</td>
  <td>9</td>
  </tr>
 <tr>
  <td>Cindy</td>
  <td>Victor</td>
  <td>2018</td>
  <td>0</td>
  <td>4</td>
  </tr>
<table>

> *`Sent1`: the number of emails that partner1 sent to partner2; *`Sent2`: the number of emails that partner2 sent to partner1

Currently, I come up with this:

* SDL:

```py
type person {
    required property name -> str;
};

type relation {
    required property relation_id -> int16;
    property inityear -> int16;
};

type relationship {
    # binary relation
    required link relation_instance -> relation;
    required link user -> person;
    property sent -> int16;
};
```

* insert:

```py
INSERT relation {relation_id := 1000, inityear := 2010};
INSERT relation {relation_id := 1001, inityear := 2007};
INSERT relation {relation_id := 1002, inityear := 2018};
INSERT relation {relation_id := 1003, inityear := 2018};

INSERT person {name := 'Alice'};
INSERT person {name := 'Bob'};
INSERT person {name := 'Cindy'};
INSERT person {name := 'Mike'};
INSERT person {name := 'Victor'};

INSERT relationship {
    relation_instance := (SELECT relation FILTER .relation_id = 1000 LIMIT 1),
    user := (SELECT person FILTER .name= 'Alice' LIMIT 1),
    sent := 6
};
INSERT relationship {
    relation_instance := (SELECT relation FILTER .relation_id = 1000 LIMIT 1),
    user := (SELECT person FILTER .name= 'Bob' LIMIT 1),
    sent := 7
};

...
```

* query:

```py
# edgeql
SELECT relationship {
    relation_instance: {inityear},
    user: {name},
    sent
} FILTER .relation_instance.relation_id = 1001;

# graphql
{
  relationship(filter: {relation_instance: {relation_id: {eq: 1001}}}){
    relation_instance{inityear}
    user{name}
    sent
  }
}
```

* result:

```bash
# edgeql query result
{
  default::relationship {relation_instance: default::relation {inityear: 2007}, user: default::person {name: 'Cindy'}, sent: 4},
  default::relationship {relation_instance: default::relation {inityear: 2007}, user: default::person {name: 'Alice'}, sent: 1},
}

# graphql query result
{
  "data": {
    "relationship": [
      {
        "relation_instance": {
          "inityear": 2007
        },
        "user": {
          "name": "Cindy"
        },
        "sent": 4
      },
      {
        "relation_instance": {
          "inityear": 2007
        },
        "user": {
          "name": "Alice"
        },
        "sent": 1
      }
    ]
  }
}
```

It is not ideal enough since there would exist duplicate `inityear` and the schema defined above seems not the native way of storing binary interaction. 

*I am wondering whether there is an elegant way to achieve this.* (If I miss something in the tutorial, forgive me for this 🤧).

**Comment by 1st1** [comment]
Hi!

You don't need to manually create relation ids, links take care of that for you.

Here's how I'd design your schema:

```
type person {
    required property name -> str;
    multi link partners -> person;
};
```

If you have relationship-specific data you can use [link properties](https://www.edgedb.com/docs/edgeql/expressions/shapes#link-properties) for that:

```
type person {
    required property name -> str;
    multi link partners -> person {
        property sent -> int64;
    }
};
```

and then query it like


```
select person {
   name,
   partners: person {
       name,
       @sent
   }
}
```

**Reply by NatureGeorge** [reply]
It looks handy. What if I want to retrieve properties in a bidirectional form just like its' raw data? I have been thinking about this for a while:

#### Goal: retrieve `Cindy`'s relationship and transform it into the interaction table

```py
SELECT person {
    `source` := .name,
    `target` := (
        SELECT (person.partners.name, person.partners@sent)
        FILTER person.partners.name = 'Cindy') IF .name != 'Cindy' ELSE
       (SELECT (person.partners.name, person.partners@sent))
 } FILTER EXISTS .`target`;
```

```py
{
  default::person {source: 'Alice', target: {('Cindy', 1)}},
  default::person {source: 'Cindy', target: {('Alice', 4), ('Victor', 0)}},
  default::person {source: 'Victor', target: {('Cindy', 4)}},
}
```

I can reorganize the result derived above and reproduce the raw data with it, but I am still wondering whether there is somewhere that the query language or the schema design can help a hand.

**Reply by NatureGeorge** [reply]
I got my way. Thank you!
